clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -323,6 +351,52 @@ private InetAddress clientAddress() {
         return transportLayer.socketChannel().socket().getInetAddress();
     }
 
+    private void handleSaslToken(byte[] clientToken) throws IOException {
+        if (!enableKafkaSaslAuthenticateHeaders) {
+            byte[] response = saslServer.evaluateResponse(clientToken);
+            if (response != null) {
+                netOutBuffer = new NetworkSend(connectionId, ByteBuffer.wrap(response));
+                flushNetOutBufferAndUpdateInterestOps();
+            }
+        } else {
+            ByteBuffer requestBuffer = ByteBuffer.wrap(clientToken);
+            RequestHeader header = RequestHeader.parse(requestBuffer);
+            ApiKeys apiKey = header.apiKey();
+            short version = header.apiVersion();
+            RequestContext requestContext = new RequestContext(header, connectionId, clientAddress(),
+                    KafkaPrincipal.ANONYMOUS, listenerName, securityProtocol);
+            RequestAndSize requestAndSize = requestContext.parseRequest(requestBuffer);
+            if (apiKey != ApiKeys.SASL_AUTHENTICATE) {
+                this.error = Errors.ILLEGAL_SASL_STATE;
+                IllegalSaslStateException e = new IllegalSaslStateException("Unexpected Kafka request of type " + apiKey + " during SASL authentication.");
+                sendKafkaResponse(requestContext, requestAndSize.request.getErrorResponse(e));
+                throw e;
+            }
+            if (!Protocol.apiVersionSupported(apiKey.id, version)) {
+                this.error = Errors.UNSUPPORTED_VERSION;
+                // We cannot create an error response if the request version of SaslAuthenticate is not supported
+                // This should not normally occur since clients typically check supported versions using ApiVersionsRequest
+                throw new UnsupportedVersionException("Version " + version + " is not supported for apiKey " + apiKey);
+            }
+            SaslAuthenticateRequest saslAuthenticateRequest = (SaslAuthenticateRequest) requestAndSize.request;
+
+            try {
+                byte[] responseToken = saslServer.evaluateResponse(Utils.readBytes(saslAuthenticateRequest.saslAuthBytes()));
+                // For versions without SASL_AUTHENTICATE header, send a response only if there is a response token to be sent.
+                // For versions with SASL_AUTHENTICATE header, send a response to SASL_AUTHENTICATE request even if token is empty.
+                if (responseToken != null || enableKafkaSaslAuthenticateHeaders) {
```

Comment:
It seems the if test is unnecessary since enableKafkaSaslAuthenticateHeaders is always true here?